src/System.Net.NetworkInformation/src/System/Net/NetworkInformation/SystemIPGlobalProperties.cs

```diff
@@ -0,0 +1,480 @@
+        {
+            List<IPEndPoint> list = new List<IPEndPoint>();
+            List<SystemTcpConnectionInformation> connections = GetAllTcpConnections();
+            foreach (TcpConnectionInformation connection in connections)
+            {
+                if (connection.State == TcpState.Listen)
+                {
+                    list.Add(connection.LocalEndPoint);
+                }
+            }
+
+            return list.ToArray();
+        }
+
+        /// 
+        /// Gets the active TCP connections. Uses the native GetTcpTable API.
+        private List<SystemTcpConnectionInformation> GetAllTcpConnections()
+        {
+            uint size = 0;
+            uint result = 0;
+            SafeLocalAllocHandle buffer = null;
+            List<SystemTcpConnectionInformation> tcpConnections = new List<SystemTcpConnectionInformation>();
+
+            // Check if it supports IPv4 for IPv6 only modes.
+            if (Socket.OSSupportsIPv4)
+            {
+                // Get the buffer size needed.
+                result = Interop.IpHlpApi.GetTcpTable(SafeLocalAllocHandle.Zero, ref size, true);
+
+                while (result == Interop.IpHlpApi.ERROR_INSUFFICIENT_BUFFER)
+                {
+                    try
+                    {
+                        // Allocate the buffer and get the TCP table.
+                        buffer = SafeLocalAllocHandle.LocalAlloc((int)size);
+                        result = Interop.IpHlpApi.GetTcpTable(buffer, ref size, true);
+
+                        if (result == Interop.IpHlpApi.ERROR_SUCCESS)
+                        {
+                            // The table info just gives us the number of rows.
+                            IntPtr newPtr = buffer.DangerousGetHandle();
+                            Interop.IpHlpApi.MibTcpTable tcpTableInfo = Marshal.PtrToStructure<Interop.IpHlpApi.MibTcpTable>(newPtr);
+
+                            if (tcpTableInfo.numberOfEntries > 0)
+                            {
+                                // Skip over the tableinfo to get the inline rows.
+                                newPtr = (IntPtr)((long)newPtr + Marshal.SizeOf(tcpTableInfo.numberOfEntries));
+
+                                for (int i = 0; i < tcpTableInfo.numberOfEntries; i++)
+                                {
+                                    Interop.IpHlpApi.MibTcpRow tcpRow = Marshal.PtrToStructure<Interop.IpHlpApi.MibTcpRow>(newPtr);
+                                    tcpConnections.Add(new SystemTcpConnectionInformation(tcpRow));
+
+                                    // Increment the pointer to the next row.
+                                    newPtr = (IntPtr)((long)newPtr + Marshal.SizeOf(tcpRow));
+                                }
+                            }
+                        }
+                    }
+                    finally
+                    {
+                        if (buffer != null)
```

Comment:
How about a using block for buffer instead of this explicit try/finally?  Same below in the IPv6 section.